File: examples/tensorly/runtests-tensorly-tucker.py
import matplotlib.pyplot as plt
import tensorly as tl
import numpy as np
from scipy.misc import face, imresize
from tensorly.decomposition import non_negative_parafac
from tensorly.decomposition import non_negative_tucker
from tensorly.decomposition import tucker
from math import ceil

from tensorly.base import tensor_to_vec, partial_tensor_to_vec
from tensorly.datasets.synthetic import gen_image
from tensorly.random import check_random_state
from tensorly.regression.kruskal_regression import KruskalRegressor
import tensorly.backend as T
import time

tl.set_backend('numpy')
rng = check_random_state(1)
X = T.tensor(rng.normal(size=(1000, 1000, 1000), loc=0, scale=1))
start_time = time.time()
core, tucker_factors = non_negative_tucker(X, rank=[10,10,10], init='svd', tol=10e-12, verbose=True, n_iter_max=2)
print("--- %s seconds ---" % (time.time() - start_time))

tl.set_backend('mxnet')
rng = check_random_state(1)
X = T.tensor(rng.normal(size=(1000, 1000, 1000), loc=0, scale=1))
start_time = time.time()
core, tucker_factors = non_negative_tucker(X, rank=[10,10,10], init='svd', tol=10e-12, verbose=True, n_iter_max=2)
print("--- %s seconds ---" % (time.time() - start_time))

tl.set_backend('pytorch')
rng = check_random_state(1)
X = T.tensor(rng.normal(size=(1000, 1000, 1000), loc=0, scale=1))
start_time = time.time()
core, tucker_factors = non_negative_tucker(X, rank=[10,10,10], init='svd', tol=10e-12, verbose=True, n_iter_max=2)
print("--- %s seconds ---" % (time.time() - start_time))

tl.set_backend('tensorflow')
rng = check_random_state(1)
X = T.tensor(rng.normal(size=(1000, 1000, 1000), loc=0, scale=1))
start_time = time.time()
core, tucker_factors = non_negative_tucker(X, rank=[10,10,10], init='svd', tol=10e-12, verbose=True, n_iter_max=2)
print("--- %s seconds ---" % (time.time() - start_time))

File: examples/tensorly/runtests-tensorly-tucker.jl
import NTFk
import TensorDecompositions

Random.seed!(1)
csize = (2, 3, 4)
tsize = (3, 5, 10)
Tucker_orig = NTFk.rand_tucker(csize, tsize, factors_nonneg=true, core_nonneg=true)
T_orig = TensorDecompositions.compose(Tucker_orig);

ttf = NTFk.tlanalysis(T_orig, [csize...])

t = NTFk.analysis(T_orig, [csize], 1)


import PyCall
@PyCall.pyimport tensorly as tl
@PyCall.pyimport tensorly.decomposition as td

tf = PyCall.pyimport("tensorflow")
o = tf[:random_normal]((3,3,3))
o = tf[:random_normal]((10,20))[:numpy]()
tl.set_backend("tensorflow");
t = td.non_negative_tucker(o, rank=(2,2,2), n_iter_max=2)

Random.seed!(1)
csize = (2, 2, 2)
tsize = (3, 3, 3)
Tucker_orig = NTFk.rand_tucker(csize, tsize, factors_nonneg=true, core_nonneg=true)
T_orig = TensorDecompositions.compose(Tucker_orig);

@info("numpy")
tl.set_backend("numpy")
@time o = tl.backend[:tensor](T_orig)
@time t = td.non_negative_tucker(o, rank=csize, n_iter_max=2)

@info("mxnet")
tl.set_backend("mxnet")
@time o = tl.backend[:tensor](T_orig)
@time t = td.non_negative_tucker(o, rank=csize, n_iter_max=2)

@info("pytorch")
tl.set_backend("pytorch")
@time o = tl.backend[:tensor](T_orig)
@time t = td.non_negative_tucker(o, rank=csize, n_iter_max=2)

@info("tensorflow")
tl.set_backend("tensorflow");
@time o = tl.backend[:tensor](T_orig)
@time t = td.non_negative_tucker(o, rank=csize, n_iter_max=2)

Random.seed!(1)
csize = (10, 10, 10)
tsize = (1000, 1000, 1000)
Tucker_orig = NTFk.rand_tucker(csize, tsize, factors_nonneg=true, core_nonneg=true)
T_orig = TensorDecompositions.compose(Tucker_orig);

@info("numpy")
tl.set_backend("numpy")
@time o = tl.backend[:tensor](T_orig);
@time td.non_negative_tucker(o, rank=csize, n_iter_max=2);

@info("mxnet")
tl.set_backend("mxnet")
@time o = tl.backend[:tensor](T_orig);
@time td.non_negative_tucker(o, rank=csize, n_iter_max=2);

@info("pytorch")
tl.set_backend("pytorch")
@time o = tl.backend[:tensor](T_orig);
@time td.non_negative_tucker(o, rank=csize, n_iter_max=2);

@info("tensorflow")
tl.set_backend("tensorflow");
@time o = tl.backend[:tensor](T_orig);
@time td.non_negative_tucker(o, rank=csize, n_iter_max=2);

@info("numpy")
tl.set_backend("numpy")
@time td.non_negative_tucker(tl.backend[:tensor](T_orig), rank=csize, n_iter_max=2);

@info("mxnet")
tl.set_backend("mxnet")
@time td.non_negative_tucker(tl.backend[:tensor](T_orig), rank=csize, n_iter_max=2);

@info("pytorch")
tl.set_backend("pytorch")
@time td.non_negative_tucker(tl.backend[:tensor](T_orig), rank=csize, n_iter_max=2);

@info("tensorflow")
tl.set_backend("tensorflow");
@time td.non_negative_tucker(tl.backend[:tensor](T_orig), rank=csize, n_iter_max=2);